proposals/interface-types/working-notes/scenarios/chooser.md

```diff
@@ -0,0 +1,680 @@
+
+## Introduction
+
+Keeping track of memory that is allocated in order to allow string processing
+can be challenging. This example illustrates an extreme case that involves
+non-determinism.
+
+This scenario is based on the idea of using run-time information in order to
+control the flow of information. In particular, consider the C++ chooser
+function:
+
+```C++
+typedef std::shared_ptr<std::string> shared_string;
+
+shared_string nondeterministic_choice(shared_string one, shared_string two) {
+  return random() > 0.5 ? std::move(one) : std_move(two);
+}
+```
+
+This function takes two string arguments and returns one of them. Regardless of
+the merits of this particular function, it sets up significant challenges should
+we want to expose it using Interface Types. Specifically, there are two
+resources entering the function, with just one leaving. However, when exposed as
+an Interface Type function, all these resources must be created and properly
+disposed of within the adapter code itself.
+
+This note focuses on the techniques that enable this to be achieved reliably.
+
+## Exporting the Chooser
+
+The Interface Type function signature for our chooser is simple: the function
+takes two `string`s and returns one:
+
+```wasm
+(@interface func (export "chooser")
+  (param $left string)
+  (param $right string)
+  (result string)
+  ...
+)
+```
+
+One of the specific challenges in this scenario is the handling of shared
+pointers. We 'require' that the core `nondeterministic_choice` function honor
+the semantics of proper reference counting the input arguments and the returned
+string.
+
+However, there are actually _two_ schemes in play in this scenario: the
+Interface Types management of resources and the C++ implementation of
+`shared_ptr`. Effectively, in addition to lifting and lowering the `string`
+values we must also lift and lower the ownership between Interface Types and
+core WASM.
+
+In this scenario we take a somewhat simplified view of C++'s implementation of
+shared pointers: a shared pointer is implemented as a pair consisting of a
+reference count and a raw pointer to the shared resource.
+
+>Note: In practice, the C++ implementation of shared pointers is somewhat more
+>complex; for reasons that are not important here.
+
+This results in _two_ memory allocations for the shared resource: one for the
+resource itself and one for the pointer structure -- which contains a reference
+count and a raw pointer to the resource.
+
+
+```wasm
+(@interface func $stralloc-x
+  (param $str string)
+  (result owned<i32>)        ;; return an owned shared ptr of the string
+  local.get $str
+  string.size
+  call-export "malloc-x"     ;; create string entity of right size
+  own (i32)
+    call-export "free-x"     ;; temporary, until safely re-owned
+  end
+  let (local $ptr owned<i32>) (result owned<i32>)
+    local.get $str
+    local.get $ptr
+	owned.access             ;; access the owned pointer
```

Comment:
is this indented deliberately? It seems like, but I don't understand the pattern.

Reply:
FWIW, +1.  I think a better scheme for handling when the linear form becomes obtuse is switching to the [folded](https://webassembly.github.io/spec/core/text/instructions.html#text-foldedinstr) form.  Notice that you can fold only just the parts you want and otherwise keep things in sequence.  Otherwise, I wouldn't invent a custom indention scheme.  (FWIW, in email threads, I've been wondering whether the indention was an unintended email artifact :P)

Reply:
That particular indentation was an unexpected consequence of using emacs :)
I may try the folded form in a future revision

Reply:
parenthetical note: the folded form relates to expression evaluation.
I was using indentation to reflect scope structure and other block structuring; not the same